dotnet/src/Extensions/TemplateEngine.PromptTemplateEngine/Blocks/NamedArgBlock.cs

```diff
@@ -0,0 +1,182 @@
+    /// </summary>
+    /// <param name="variables">Variables to use for rendering the named argument value when the value is a <see cref="VarBlock"/>.</param>
+    /// <returns></returns>
+    internal string GetValue(ContextVariables? variables)
+    {
+        var valueIsValidValBlock = this._valBlock != null && this._valBlock.IsValid(out var errorMessage);
+        if (valueIsValidValBlock)
+        {
+            return this._valBlock!.Render(variables);
+        }
+
+        var valueIsValidVarBlock = this._argValueAsVarBlock != null && this._argValueAsVarBlock.IsValid(out var errorMessage2);
+        if (valueIsValidVarBlock)
+        {
+            return this._argValueAsVarBlock!.Render(variables);
+        }
+
+        return string.Empty;
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="NamedArgBlock"/> class.
+    /// </summary>
+    /// <param name="text">Raw text parsed from the prompt template.</param>
+    /// <param name="logger">The <see cref="ILoggerFactory"/> to use for logging. If null, no logging will be performed.</param>
+    /// <exception cref="SKException"></exception>
+    public NamedArgBlock(string? text, ILoggerFactory? logger = null)
+        : base(NamedArgBlock.TrimWhitespace(text), logger)
+    {
+        var argParts = this.Content.Split(Symbols.NamedArgBlockSeparator);
+        if (argParts.Length != 2)
+        {
+            this.Logger.LogError("Invalid named argument `{Text}`", text);
+            throw new SKException($"A function named argument must contain a name and value separated by a '{Symbols.NamedArgBlockSeparator}' character.");
+        }
+
+        this.Name = argParts[0];
+        this._argNameAsVarBlock = new VarBlock($"{Symbols.VarPrefix}{argParts[0]}");
+        var argValue = argParts[1];
+        if (argValue.Length == 0)
+        {
+            this.Logger.LogError("Invalid named argument `{Text}`", text);
+            throw new SKException($"A function named argument must contain a quoted value or variable after the '{Symbols.NamedArgBlockSeparator}' character.");
+        }
+
+        if (argValue[0] == Symbols.VarPrefix)
+        {
+            this._argValueAsVarBlock = new VarBlock(argValue);
+        }
+        else
+        {
+            this._valBlock = new ValBlock(argValue);
+        }
+
+        return;
+    }
+
+    /// <summary>
+    /// Renders the named arg block.
+    /// </summary>
+    /// <param name="variables"></param>
+    /// <returns></returns>
+    public string Render(ContextVariables? variables)
+    {
+        return this.Content;
+    }
+
+    /// <summary>
+    /// Returns whether the named arg block has valid syntax.
+    /// </summary>
+    /// <param name="errorMsg">An error message that gets set when the named arg block is not valid.</param>
+    /// <returns></returns>
+#pragma warning disable CA2254 // error strings are used also internally, not just for logging
+    public override bool IsValid(out string errorMsg)
```

Comment:
I found an issue in this method. I tested it using your example from unit test, but I extended it a little bit:
`{{my.sayAge name=\"Mario\" birthdate=$birthdate exclamation='Wow, that\\'s surprising' test=5}}`, so I added `test=5` parameter. 

Probably out of scope of this PR, but it looks like passing plain numbers is not supported at the moment, which is a little bit confusing. I'm wondering if it makes sense to add this functionality in the future. Maybe it worth to create separate issue for that and investigate it further.

The problem in this method - if I use `test=5` syntax, which is invalid at the moment, the value check will fail in line 114 and there will be some information in `errorMsg` out variable. Then, in line 128 we are performing name check, and in my case the name is valid, but it uses the same `errorMsg` out variable, which means that this variable will be overwritten, and as a result - empty. In this case, method `RenderAsync` will throw `SKException` without any information:
![image](https://github.com/microsoft/semantic-kernel/assets/13853051/62153f29-80de-4f5c-b5bb-9e64502064ce)

Overall, I think we should avoid passing `errorMsg` as out variable through the stack trace, because it's hard to maintain its state. I'm wondering, if we can just immediately throw an exception in case of invalid syntax with error message. This will resolve the problem which I described above and potential similar problems in the future.

As part of this PR, it would be great to add more unit tests, specifically with invalid scenarios and verify the expected behavior.

Reply:
Discussed offline - here's the summary:

* Keep or remove the out variable for `IsValid` a method on the _base_ class for NamedArgBlock?
  * Outcome: keep. Re-evaluate as a separate issue.
  * Reason: This would change the interface for many other classes, and I don't fully understand the original intent behind it
* How to avoid confusing blank exceptions?
  * Outcome: throw immediately after encountering an issue.
  * Reason: Simplest solution, less error prone, doesn't require API change.
*  Add more unit tests for valid and invalid scenarios
* To support numbers/unquoted arg values?
  * Currently working as designed (see ADR). Only quoted and variable values are supported.
  * I'm open to re-evaluating this as a separate issue but wanted to keep things simple.
  
The reason I came to this decision had to do with this line from the ADR for supporting multiple args in native functions:

> Added support for non-string arguments. All C# primitive types and many core .NET types are supported, with their corresponding TypeConverters used to parse the string context variable into the appropriate type. Custom types attributed with TypeConverterAttribute may also be used, and the associated TypeConverter will be used as is appropriate.

So the args are expected to be saved as strings in context anyways even if the value looks like an int in the template. However, it is still possible for the engine to be flexible enough to save all non-variable values as strings to context.

Reply:
Added some more tests for the error message in my latest commit. I already had a test for unquoted/non-variable values, so I updated that to check the error message.